### PR TITLE
IsFinite methods for min-plus and max-plus matrix semigroups

### PR DESCRIPTION
**Pull request by stuartburrell**
The following has tests, documentation and implementations for:

* `IsFinite` for max-plus and min-plus matrix semigroups
* `IsTorsion` for max-plus matrix semigroups
* `NormalizeSemigroup` for max-plus matrix semigroup

and

* `SpectralRadius` for max-plus matrices
* `InverseOp` for max-plus matrices
* `UnweightedPrecedenceDigraph` for max-plus matrices
* `RadialEigenvector` for max-plus matrices

### CODE REVIEW

**Comment by james-d-mitchell** [comment]
I reformatted the pull request text a wee bit @stuartburrell, but before we can go any further you'll have to fix the conflicts mention above.

**Comment by james-d-mitchell** [comment]
@stuartburrell All of your commits are duplicated, could you please fix that too?

File: doc/maxplusmat.xml [hunk 1]
```diff
@@ -197,3 +197,79 @@ gap> PeriodNTPMatrix(mat);
...
+<#/GAPDoc>
+
+<#GAPDoc Label="InverseOp">
+  <ManSection>
+    <Oper Name = "InverseOp" Arg = "mat"/>
+    <Returns>A max-plus matrix.</Returns>
+    <Description>
+      If <A>mat</A> is an invertible max-plus matrix (i.e. belongs to the
+      category <Ref Filt = "IsMaxPlusMatrix"/>) and is a row permutation
+      applied to the identity), then <C>InverseOp</C> returns the inverse
```
Comment (james-d-mitchell) [inline doc/maxplusmat.xml]: There's and extra bracket in this line `identity)`

File: doc/maxplusmat.xml [hunk 2]
```diff
@@ -197,3 +197,79 @@ gap> PeriodNTPMatrix(mat);
...
+    </Description>
+  </ManSection>
+<#/GAPDoc>
+
+<#GAPDoc Label="RadialEigenvector">
+  <ManSection>
+    <Oper Name = "RadialEigenvector" Arg = "mat"/>
+    <Returns>A vector.</Returns>
+    <Description>
+      If <A>mat</A> is an <K>n</K> by <K>n</K> max-plus matrix (i.e. belongs to
```
Comment (james-d-mitchell) [inline doc/maxplusmat.xml]: The `<K>` tags are for keywords, you should use `<C>` tags here instead.

File: doc/maxplusmat.xml [hunk 3]
```diff
@@ -197,3 +197,79 @@ gap> PeriodNTPMatrix(mat);
...
+<#/GAPDoc>
+
+<#GAPDoc Label="UnweightedPrecedenceDigraph">
+  <ManSection>
+    <Oper Name = "UnweightedPrecedenceDigraph" Arg = "mat"/>
+    <Returns>A digraph.</Returns>
+    <Description>
+      If <A>mat</A> is a max-plus matrix (i.e. belongs to the category <Ref
+      Filt = "IsMaxPlusMatrix"/>), then <C>UnweightedPrecedenceDigraph</C>
+      returns the unweighted precedence digraph corresponding to <K>mat</K>.
```
Comment (james-d-mitchell) [inline doc/maxplusmat.xml]: Wrong tags around `<K>mat</K>` should be `<A>`

File: doc/maxplusmat.xml [hunk 4]
```diff
@@ -197,3 +197,79 @@ gap> PeriodNTPMatrix(mat);
...
+      
+      For an <K>n</K> by <K>n</K> matrix <K>mat</K>, the unweighted precedence
+      digraph is defined as the digraph on <K>n</K> points where vertex <K>i</K>
+      is adjacent to vertex <K>j</K> if and only if <K>mat[i,j]</K> is not equal
```
Comment (james-d-mitchell) [inline doc/maxplusmat.xml]: Wrong tags here too, `<K>` -> `<C>` or `<A>` for `mat`

File: doc/semimaxplus.xml [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  <ManSection>
+    <Attr Name = "IsFinite" Arg = "S"/>
+    <Returns><K>true</K> or <K>false</K>. </Returns>
+    <Description>
+      If <A>S</A> is a max-plus or min-plus matrix semigroup (i.e. belongs to
+      the category <Ref Filt = "IsMaxPlusMatrixSemigroup"/>), then
+      <C>IsFinite</C> returns <K>true</K> if <A>S</A> is finite and 
+      <K>false</K> otherwise. This method is based on
+      <Cite Key="Gaubert1996aa"/> (max-plus) and <Cite Key="Simon1978aa"/>
+      (min-plus). Note, in the min-plus case, all entries must be non-negative.
```
Comment (james-d-mitchell) [inline doc/semimaxplus.xml]: What does this mean exactly that all entries must be non-negative? That for the method to work the entries have to be non-negative, or for the semigroup to be finite? 

Could you add a sentence that says something to the effect of "this method does not involve enumerating the semigroup."

File: doc/semimaxplus.xml [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  <ManSection>
+    <Oper Name = "NormalizeSemigroup" Arg = "S"/>
+    <Returns>A semigroup.</Returns>
+    <Description>
+      This method applies to max-plus matrix semigroups (i.e. those belonging to
+      the category <Ref Filt = "IsMaxPlusMatrixSemigroup"/>) that are 
+      finitely generated, such that the spectral radius of the matrix equal to
+      the sum of the generators (with respect to the max-plus semiring) is zero. 
+      <C>NormalizeSemigroup</C> returns  a semigroup of matrices all with 
+      stricly non-positive entries. Note that the output is isomorphic to a
```
Comment (james-d-mitchell) [inline doc/semimaxplus.xml]: `stricly`

**Comment by james-d-mitchell** [comment]
This looks pretty good. I made some inline comments (minor typos mostly), which it would be great if you could fix.

**Comment by james-d-mitchell** [comment]
Also I made a pull request for your branch which contains some minor code formatting changes. Please take a look.

**Comment by stuartburrell** [comment]
Good morning.

Thanks for all the comments!

I'll do the fixes and commit as soon as possible.

Cheers,

Stuart

**Pull request merged**